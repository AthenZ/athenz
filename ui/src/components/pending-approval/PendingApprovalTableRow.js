/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { colors } from '../denali/styles/colors';
import styled from '@emotion/styled';

import CheckBox from '../denali/CheckBox';
import TextArea from '../denali/TextArea';
import FlatPicker from '../flatpicker/FlatPicker';
import Icon from '../denali/icons/Icon';
import DateUtils from '../utils/DateUtils';
import NameUtils from '../utils/NameUtils';

const TableTd = styled.td`
    text-align: left;
    vertical-align: middle;
    word-break: break-all;
    padding: 5px 0px 5px 15px;
    white-space: nowrap;
`;

const StyledText = styled.p`
    color: #3570f4;
`;

const TableTdText = styled.td`
    text-align: left;
    vertical-align: middle;
    padding: 5px 0px 5px 15px;
`;

const ErrorDiv = styled.div`
    color: #ea0000;
`;

const TableTdDomain = styled.td`
    text-align: left;
    vertical-align: middle;
    word-break: break-all;
    padding: 5px 0px 5px 0px;
`;

const FlatPickrInputDiv = styled.div`
    & > div input {
        position: relative;
        font: 300 14px HelveticaNeue-Reg, Helvetica, Arial, sans-serif;
        background-color: ${(props) =>
            props.disabled ? colors.grey500 : 'rgba(53, 112, 244, 0.05)'};
        box-shadow: none;
        color: rgb(48, 48, 48);
        height: 16px;
        min-width: 50px;
        text-align: left;
        border-width: 2px;
        border-style: solid;
        border-color: transparent;
        border-image: initial;
        border-radius: 2px;
        flex: 1 0 auto;
        margin: 0px;
        margin-top: 5px;
        outline: none;
        padding: 0.6em 12px;
        transition: background-color 0.2s ease-in-out 0s,
            color 0.2s ease-in-out 0s, border 0.2s ease-in-out 0s;
        width: 10em;
    }
`;

const StyledTextArea = styled(TextArea)`
    & > textarea {
        resize: vertical;
    }
`;

const ApproveTd = styled.td`
    vertical-align: middle;
    word-break: break-all;
    text-align: center;
    position: absolute;
    width: 11em;
    right: 11em;
    height: 96px;
    background-color: ${(props) => props.color};
    padding: 5px 0px 5px 14px;
`;

const RejectTd = styled.td`
    vertical-align: middle;
    word-break: break-all;
    text-align: center;
    border-right: none;
    position: absolute;
    width: 11em;
    right: 0em;
    height: 96px;
    background-color: ${(props) => props.color};
    padding: 5px 0px 5px 0px;
`;

const IconDiv = styled.div`
    margin-top: 45px;
`;

const TableRow = styled.tr`
    background-color: ${(props) => props.color};
`;

export default class PendingApprovalTableRow extends React.Component {
    constructor(props) {
        super(props);
        this.dateUtils = new DateUtils();
    }

    render() {
        const key =
            this.props.domainName + this.props.memberName + this.props.roleName;
        const fpExpiryKey = NameUtils.getFlatPickrKey(key + 'expiry');
        const fpReviewReminderKey = NameUtils.getFlatPickrKey(
            key + 'reviewReminder'
        );
        let opaqueBackground = 'rgba(255, 255, 255, 1)';
        if (colors.row === this.props.color) {
            opaqueBackground = 'rgba(244, 248, 254, 1)';
        }
        let approveColor = this.props.checked
            ? colors.grey500
            : colors.green600;
        let rejectColor = this.props.checked
            ? colors.grey500
            : colors.statusDanger;
        let approveOnClick = this.props.checked
            ? () => {}
            : this.props.pendingDecision.bind(this, key, true);
        let rejectOnClick = this.props.checked
            ? () => {}
            : this.props.pendingDecision.bind(this, key, false);
        return (
            <TableRow
                color={this.props.color}
                data-testid='pending-approval-table-row'
                key={key}
            >
                <TableTd>
                    <CheckBox
                        checked={this.props.checked}
                        name='checkbox-row'
                        id={
                            this.props.domainName +
                            this.props.memberName +
                            this.props.roleName
                        }
                        onChange={() => {
                            this.props.checkedClick(key);
                        }}
                    />
                </TableTd>
                <TableTdDomain>{this.props.domainName}</TableTdDomain>
                <TableTd>{this.props.category}</TableTd>
                <TableTd>{this.props.roleName}</TableTd>
                <TableTdText>
                    <p>{this.props.memberNameFull}</p>
                    <p>
                        {'('}
                        {this.props.memberName}
                        {')'}
                    </p>
                </TableTdText>
                <TableTdText colSpan={2}>{this.props.userComment}</TableTdText>
                <TableTdText>
                    <p>{this.props.requestPrincipalFull}</p>
                    <p>
                        {'('}
                        {this.props.requestPrincipal}
                        {')'}
                    </p>
                </TableTdText>
                <TableTd>
                    {this.dateUtils.getLocalDate(
                        this.props.requestTime,
                        'UTC',
                        'UTC'
                    )}
                </TableTd>
                <TableTd>
                    <StyledTextArea
                        id={key}
                        disabled={this.props.checked}
                        placeholder='Justification'
                        rows='4'
                        width='100%'
                        onChange={(event) => {
                            this.props.auditRefChange(key, event);
                        }}
                    />
                    {this.props.auditRefMissing ? (
                        <ErrorDiv>Justification is Required</ErrorDiv>
                    ) : null}
                </TableTd>
                <TableTd>
                    {this.props.category !== 'group' && (
                        <FlatPickrInputDiv disabled={this.props.checked}>
                            <FlatPicker
                                id={fpExpiryKey}
                                onChange={(date) => {
                                    this.props.dateChange(key, date, 'expiry');
                                }}
                                disabled={this.props.checked}
                                clearExpiry={this.props.clearExpiry}
                                value={this.props.requestedExpiry}
                                nomargin={true}
                            />
                        </FlatPickrInputDiv>
                    )}
                </TableTd>
                <TableTd>
                    {this.props.category !== 'group' && (
                        <FlatPickrInputDiv disabled={this.props.checked}>
                            <FlatPicker
                                id={fpReviewReminderKey}
                                onChange={(date) => {
                                    this.props.dateChange(
                                        key,
                                        date,
                                        'reviewReminder'
                                    );
                                }}
                                placeholder={'Reminder (Optional)'}
                                disabled={this.props.checked}
                                clearReviewReminder={this.props.reviewReminder}
                                value={this.props.requestedReviewReminder}
                                nomargin={true}
                            />
                        </FlatPickrInputDiv>
                    )}
                </TableTd>
                <ApproveTd color={opaqueBackground}>
                    <IconDiv>
                        <Icon
                            disabled={this.props.checked}
                            icon={'check-circle'}
                            onClick={approveOnClick}
                            color={approveColor}
                            isLink
                            size={'1.25em'}
                            verticalAlign={'text-bottom'}
                        />
                    </IconDiv>
                </ApproveTd>
                <RejectTd color={opaqueBackground}>
                    <IconDiv>
                        <Icon
                            disabled={this.props.checked}
                            icon={'decline'}
                            onClick={rejectOnClick}
                            color={rejectColor}
                            isLink
                            size={'1.25em'}
                            verticalAlign={'text-bottom'}
                        />
                    </IconDiv>
                </RejectTd>
            </TableRow>
        );
    }
}
