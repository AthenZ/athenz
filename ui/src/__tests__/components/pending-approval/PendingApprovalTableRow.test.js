/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import PendingApprovalTableRow from '../../../components/pending-approval/PendingApprovalTableRow';

describe('PendingApprovalTable', () => {
    it('should render', () => {
        const color = 'white';
        const domainName = 'a';
        const memberName = 'b';
        const roleName = 'c';
        const checked = false;
        const userComment = 'd';

        const { getByTestId } = render(
            <table>
                <tbody>
                    <PendingApprovalTableRow
                        color={color}
                        domainName={domainName}
                        memberName={memberName}
                        roleName={roleName}
                        checkedf={checked}
                        userComment={userComment}
                        pendingDecision={() => {}}
                        auditRefMissing={false}
                    />
                </tbody>
            </table>
        );

        const pendingapprovaltablerow = getByTestId(
            'pending-approval-table-row'
        );
        expect(pendingapprovaltablerow).toMatchSnapshot();
    });
});
