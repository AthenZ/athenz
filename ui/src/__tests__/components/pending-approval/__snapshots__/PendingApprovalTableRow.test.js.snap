// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingApprovalTable should render 1`] = `
<tr
  class="css-1sje1x9"
  color="white"
  data-testid="pending-approval-table-row"
>
  <td
    class="css-t52g8d"
  >
    <div
      class="css-1xv7yfr-checkbox denali-checkbox"
      data-testid="checkbox-wrapper"
    >
      <input
        data-partial="false"
        id="abc"
        name="checkbox-row"
        type="checkbox"
      />
      <label
        for="abc"
      />
    </div>
  </td>
  <td
    class="css-frhcp6"
  >
    a
  </td>
  <td
    class="css-t52g8d"
  >
    c
  </td>
  <td
    class="css-lszrs5"
  >
    <p />
    <p>
      (
      b
      )
    </p>
  </td>
  <td
    class="css-lszrs5"
    colspan="2"
  >
    d
  </td>
  <td
    class="css-lszrs5"
  >
    <p />
    <p>
      (
      )
    </p>
  </td>
  <td
    class="css-t52g8d"
  >
    Invalid date
  </td>
  <td
    class="css-t52g8d"
  >
    <div
      class="css-1n5z8r6-textarea denali-textarea css-6s8cn2"
      data-testid="textarea-wrapper"
    >
      <textarea
        data-testid="textarea"
        id="abc"
        placeholder="Justification"
        rows="4"
      />
    </div>
  </td>
  <td
    class="css-t52g8d"
  >
    <div
      class="css-1nca0yn"
    >
      <div
        class="fp-abc"
      >
        <input
          class="flatpickr-input"
          data-input=""
          data-testid="flatPicker"
          placeholder="Expiration (Optional)"
          type="hidden"
          value=""
        />
        <input
          class=" form-control input"
          placeholder="Expiration (Optional)"
          readonly="readonly"
          tabindex="0"
          type="text"
        />
        <a
          data-clear=""
        >
          <svg
            class="css-u87f64"
            data-testid="icon"
            height="1.25em"
            viewBox="0 0 1024 1024"
            width="1.25em"
          >
            <title>
              close
            </title>
            <path
              d="M571.733 512l226.133-224c7.901-8.071 12.777-19.131 12.777-31.33 0-9.87-3.192-18.994-8.599-26.397l0.088 0.127c-7.382-10.88-19.691-17.934-33.649-17.934-12.060 0-22.89 5.267-30.315 13.626l-0.036 0.042-226.133 226.133-226.133-226.133c-7.461-8.401-18.291-13.668-30.351-13.668-13.957 0-26.267 7.055-33.558 17.793l-0.091 0.142c-5.319 7.276-8.51 16.4-8.51 26.27 0 12.199 4.876 23.259 12.785 31.338l-0.008-0.008 226.133 224-226.133 224c-7.901 8.071-12.777 19.131-12.777 31.33 0 9.87 3.192 18.994 8.599 26.397l-0.088-0.127c7.382 10.88 19.691 17.934 33.649 17.934 12.060 0 22.89-5.267 30.315-13.626l0.036-0.042 226.133-226.133 226.133 226.133c7.461 8.401 18.291 13.668 30.351 13.668 13.957 0 26.267-7.055 33.558-17.793l0.091-0.142c5.319-7.276 8.51-16.4 8.51-26.27 0-12.199-4.876-23.259-12.785-31.338l0.008 0.008z"
            />
          </svg>
        </a>
      </div>
    </div>
  </td>
  <td
    class="css-1bcsa8k"
  >
    <svg
      class="css-fw471q"
      data-testid="icon"
      height="1.25em"
      viewBox="0 0 1024 1024"
      width="1.25em"
    >
      <title>
        check-circle
      </title>
      <path
        d="M512 128c212.077 0 384 171.923 384 384s-171.923 384-384 384c-212.077 0-384-171.923-384-384v0c0-212.077 171.923-384 384-384v0zM512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0-259.206-210.128-469.333-469.333-469.333v0z"
      />
      <path
        d="M396.16 691.413l-126.293-126.293c-8.506-8.066-13.798-19.448-13.798-32.066 0-9.118 2.763-17.59 7.498-24.625l-0.1 0.158c7.83-10.919 20.483-17.949 34.778-17.949 11.841 0 22.556 4.824 30.286 12.613l0.003 0.003 98.133 97.707 268.373-268.373c7.549-6.884 17.635-11.1 28.706-11.1 23.564 0 42.667 19.103 42.667 42.667 0 11.017-4.176 21.060-11.031 28.63l0.032-0.036-298.667 298.667c-7.733 7.795-18.45 12.621-30.293 12.621s-22.56-4.826-30.291-12.618l-0.003-0.003z"
      />
    </svg>
  </td>
  <td
    class="css-1cg9aph"
  >
    <svg
      class="css-1pgf9ds"
      data-testid="icon"
      height="1.25em"
      viewBox="0 0 1024 1024"
      width="1.25em"
    >
      <title>
        decline
      </title>
      <path
        d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM130.133 467.2c22.95-192.169 184.962-339.729 381.44-339.729 90.864 0 174.356 31.559 240.114 84.314l-0.754-0.585-539.733 539.733c-52.076-65.227-83.563-148.885-83.563-239.896 0-15.446 0.907-30.681 2.671-45.653l-0.174 1.816zM554.667 893.867c-12.252 1.372-26.457 2.154-40.844 2.154-92.264 0-177.012-32.182-243.63-85.933l0.741 0.579 539.733-537.6c53.445 64.395 85.876 147.901 85.876 238.983 0 15.772-0.972 31.316-2.861 46.577l0.184-1.826c-22.136 177.253-161.672 316.060-337.376 336.891l-1.824 0.176z"
      />
    </svg>
  </td>
</tr>
`;
